### ISSUE TITLE
Crash with long commands using mpr/big images on Debian 11

### ISSUE THREAD

**Issue by jackyhallyday** [ImageMagick/ImageMagick#6628]
### ImageMagick version

7.1.1-15 Q16-HDRI x86_64 98eceff6a:20230729

### Operating system

Linux

### Operating system, version and so on

Debian 11

### Description

Note: ImageMagick is installed by IMEI (on Github).

I am using some pretty long commands to generate multiple formats in one command. I already used this type of commands on my previous server (Debian 10), then I got a new machine with Debian 11 (and maybe a new ImageMagick version I'm not sure).
The problem is that now, some of these commands with big images are causing a server crash. The only solution is to reboot, so it's pretty serious.

### Steps to Reproduce

Here is an example. It's only exporting WebP images because I wanted to test if the optimisations for JPEG and PNG in 7.1.x could be the cause — they are not, it crashes anyways.

The src dir and its two files, plus the (empty) output dir are in the joint zip file.

```
magick convert  -define jpeg:size=2700x1800  './src/a--crash.jpg' \
-profile ./src/sRGB2014.icc -strip \
-write mpr:TMP +delete -respect-parentheses  \( mpr:TMP -resize 1350x900\> -quality 68 +write './output/a--crash-big_2x.webp' \) \
\( mpr:TMP -resize 675x450\> -quality 85 +write './output/a--crash-big.webp' \) \
\( mpr:TMP -resize 600x400\> -quality 68 -define webp:use-sharp-yuv=true +write './output/a--crash-med_2x.webp' \) \
\( mpr:TMP -resize 450x300\> -gravity center -crop 300x300+0+0 +repage -quality 75 -define webp:use-sharp-yuv=true +write './output/a--crash-small_2x.webp' \) \
\( mpr:TMP -resize 300x200\> -quality 85 -define webp:use-sharp-yuv=true +write './output/a--crash-med.webp' \) \
\( mpr:TMP -resize 225x150\> -gravity center -crop 150x150+0+0 +repage -quality 85 -define webp:use-sharp-yuv=true +write './output/a--crash-small.webp' \) \
null:
```

### Images

[Archive.zip](https://github.com/ImageMagick/ImageMagick/files/12554320/Archive.zip)

**Comment by fmw42** [comment]
Can you format your command in multiple lines to make it easier to read?

Do you have the latest webp on your new system?

What do you get from 

`magick -version`

Note: On IM 7, use magick, not magick convert.

Note 2: If you strip after setting a profile, you will remove that profile

**Comment by jackyhallyday** [comment]
I had the Debian one (0.6.1) I manually installed the latest one but it still crashes. It was pretty much expected, since it's crashing if I'm only converting to JPG too.

magic -version returns:

Version: ImageMagick 7.1.1-15 Q16-HDRI x86_64 98eceff6a:20230729 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): bzlib cairo djvu fftw fontconfig freetype gslib gvc heic jbig jng jp2 jpeg jxl lcms lqr ltdl lzma openexr pangocairo png ps raqm raw rsvg tiff webp wmf x xml zip zlib
Compiler: gcc (10.2)

Note 2: The purpose of this part is to be sure that the image is in the sRGB color space, then strip the ICC profile since it's the default profile for web browsers. Otherwise you can have a 40x40px thumbnail that weight more than 1MB due to a big CYMK profile, for instance.

**Comment by snibgo** [comment]
Each of your "`\( ... \)`" lines eats memory, because you create a new image that isn't used after that line. Not a huge amount of memory, and I doubt it would cause problems. I suggest you insert "`+delete`" immediately after each write.

**Comment by jackyhallyday** [comment]
I just upgraded to Debian 12 and it solved my problem. No more crashes 🥳 🎉  !!!

@snibgo Thanks for the advice.

**Comment by jackyhallyday** [comment]
.... but in fact, it still crashes :-(((

**Comment by jackyhallyday** [comment]
> 

How should I do it ?
I thought I was just reusing mpr:TMP (created once at the beginning) to generate new images.
How should I write my command to avoid wasting memory?
Where should I add the +delete arguments?

**Comment by fmw42** [comment]
Just before each closing parenthesis

**Comment by urban-warrior** [comment]
We're using ImageMagick 7.1.1-16 beta and your command completed without complaint over 100 iterations.  Can you try this command-set:

```
$ git clone https://github.com/ImageMagick/ImageMagick.git ImageMagick-7
$ cd ImageMagick-7
$ ./configure --disable-shared
$ make
$ mv ../src ../output .
$ ./utilities/magick convert  -define jpeg:size=2700x1800  './src/a--crash.jpg' \
-profile ./src/sRGB2014.icc -strip \
-write mpr:TMP +delete -respect-parentheses  \( mpr:TMP -resize 1350x900\> -quality 68 +write './output/a--crash-big_2x.webp' \) \
\( mpr:TMP -resize 675x450\> -quality 85 +write './output/a--crash-big.webp' \) \
\( mpr:TMP -resize 600x400\> -quality 68 -define webp:use-sharp-yuv=true +write './output/a--crash-med_2x.webp' \) \
\( mpr:TMP -resize 450x300\> -gravity center -crop 300x300+0+0 +repage -quality 75 -define webp:use-sharp-yuv=true +write './output/a--crash-small_2x.webp' \) \
\( mpr:TMP -resize 300x200\> -quality 85 -define webp:use-sharp-yuv=true +write './output/a--crash-med.webp' \) \
\( mpr:TMP -resize 225x150\> -gravity center -crop 150x150+0+0 +repage -quality 85 -define webp:use-sharp-yuv=true +write './output/a--crash-small.webp' \) \
null:
```

Did that work for you?  If not, try

```
$ gdb ./utilities/magick
(gdb) run convert  -define jpeg:size=2700x1800  './src/a--crash.jpg' \
-profile ./src/sRGB2014.icc -strip \
-write mpr:TMP +delete -respect-parentheses  \( mpr:TMP -resize 1350x900\> -quality 68 +write './output/a--crash-big_2x.webp' \) \
\( mpr:TMP -resize 675x450\> -quality 85 +write './output/a--crash-big.webp' \) \
\( mpr:TMP -resize 600x400\> -quality 68 -define webp:use-sharp-yuv=true +write './output/a--crash-med_2x.webp' \) \
\( mpr:TMP -resize 450x300\> -gravity center -crop 300x300+0+0 +repage -quality 75 -define webp:use-sharp-yuv=true +write './output/a--crash-small_2x.webp' \) \
\( mpr:TMP -resize 300x200\> -quality 85 -define webp:use-sharp-yuv=true +write './output/a--crash-med.webp' \) \
\( mpr:TMP -resize 225x150\> -gravity center -crop 150x150+0+0 +repage -quality 85 -define webp:use-sharp-yuv=true +write './output/a--crash-small.webp' \) \
null:
(gdb) bt
```

and post the backtrace here so we can see where the problem occurs.

**Comment by snibgo** [comment]
Regarding `+delete` etc:

Your command contains:
```
\( mpr:TMP -resize 1350x900\> -quality 68 +write './output/a--crash-big_2x.webp' \) \
```
That creates a new version of the mpr:TMP image. That version has roughly 1350x900 pixels. It is written to a file. It remains in IM's memory until the end of the command. That image is never used in your command after that time, so you can remove it from memory, like this:
```
\( mpr:TMP -resize 1350x900\> -quality 68 +write './output/a--crash-big_2x.webp' +delete \) \
```
This doesn't delete the created webp file, or mpr:TMP. It just removes the 1350x900 version of the image.

A million pixels is nothing, but you have six of these, and maybe more in the future, and maybe larger in the future. It is good to remove images from memory when you can.

**Comment by jackyhallyday** [comment]
> We're using ImageMagick 7.1.1-16 beta and your command completed without complaint over 100 iterations.

I haven't been able to reproduce it today, too. It crashed the whole system every time, and yesterday I had another case that crashed the system but it was a random test and I haven't been able to reproduce it.

I knew some other cases that crashed the server every time for the last three three months which led to dozens of reboots, and today everything worked fine.
Two things led to crashes: the size of the picture, and the number of images generated.
I used an image resized at 2500x1667 and 3000x2000 with Apple Preview as a source. The 2500x version worked, the 3000x version crashed.

My example in the first message would crash every time, but not if I generated fewer images.

Since it was related to the size or number of pictures generated, I made an extensive test of my server's RAM, but it didn't crash.

In conclusion, I have no idea why I had this bug, and why I had that last crash yesterday. I won't try your commands unless I have another crash and I'm closing this bug, hopefully forever.

I'm sorry that I wasted your time. Thanks a lot for your reply.

**Comment by jackyhallyday** [comment]
> A million pixels is nothing, but you have six of these, and maybe more in the future, and maybe larger in the future. It is good to remove images from memory when you can.

Sure. Thank you, I implemented it immediately.

**Comment by urban-warrior** [comment]
Try this as well, reduce your memory footprint by moving the pixel cache from memory to disk and move temporary files away from /tmp.  Add this to your CLI:

```
-limit memory 2GiB -limit map 4GiB -define registry:temporary-path=./
```

**Comment by jackyhallyday** [comment]
As I mentioned before, I suspected that the crash came from the RAM. Maybe I was right, and I *think* upgrading to Debian 12 solved the problem.
Then I had new crashes, I installed the beta version of ImageMagick as suggested (https://github.com/ImageMagick/ImageMagick/issues/6628#issuecomment-1712335459), and it crashed too.

But that was because I started to do more stuff: I was playing with JXL - and it didn't crash at first.

I'm generating multiple files at once, and it started crashing again when I was trying to generate bigger files. So I thought it could be a memory thing (again).
Then I reduced the number of the pictures that I was generating, until it crashed even when I wanted to generate a single big JXL from a picture (without ICC).

Then I used gdb (I tried with "-limit memory 2GiB -limit map 4GiB -define registry:temporary-path=./", it crashed too).
```
(gdb) run convert './src/bigFileWithoutICC.jpg' -strip  -write mpr:TMP +delete -respect-parentheses  \( mpr:TMP -resize 1500x2000\> -quality 80 +write './output/bigFileWithoutICC.jpg.jxl' +delete \)  null:
```
Which displayed immediately:
```
[New Thread 0x7fffe4bd76c0 (LWP 9535)]
[New Thread 0x7fffe43d66c0 (LWP 9536)]
[New Thread 0x7fffe3bd56c0 (LWP 9537)]
[New Thread 0x7fffe33d46c0 (LWP 9538)]
[New Thread 0x7fffe2bd36c0 (LWP 9539)]
[New Thread 0x7fffe23d26c0 (LWP 9540)]
[New Thread 0x7fffe1bd16c0 (LWP 9541)]
[New Thread 0x7fffe13d06c0 (LWP 9542)]
[New Thread 0x7fffe0bcf6c0 (LWP 9543)]
[New Thread 0x7fffe03ce6c0 (LWP 9544)]
[New Thread 0x7fffdfbcd6c0 (LWP 9545)]
[New Thread 0x7fffe8d9f6c0 (LWP 9548)]
[New Thread 0x7fffe859e6c0 (LWP 9550)]
[New Thread 0x7fffe7d9d6c0 (LWP 9551)]
[New Thread 0x7fffe759c6c0 (LWP 9552)]
[New Thread 0x7fffe6d9b6c0 (LWP 9553)]
[New Thread 0x7fffe659a6c0 (LWP 9554)]
[New Thread 0x7fffe5d996c0 (LWP 9555)]
[New Thread 0x7fffe55986c0 (LWP 9556)]
[New Thread 0x7fffdf3cc6c0 (LWP 9557)]
[New Thread 0x7fffdebcb6c0 (LWP 9558)]
[New Thread 0x7fffde3ca6c0 (LWP 9559)]
[New Thread 0x7fffddbc96c0 (LWP 9560)]
```
And then it crashed.

My CPU is a Intel Xeon-E 2386G - 6c/12t. So after a reboot, I limited the number of threads to 1, 2, 4, 6, 8.
No more crash, bingo. Just to be sure, I removed the limit and it crashed again.

So I *MIGHT* have found what my problem was.
But I've been there already, so... stay tuned. :-)


Note : $ magick identify -list format
 JXL* JXL       rw+   JPEG XL (ISO/IEC 18181) (libjxl 0.8.2)

**Comment by urban-warrior** [comment]
Looks like a threading issue?  Perhaps a bug in libgomp as opposed to ImageMagick.    We have a  Intel(R) Xeon(R) E-2378G CPU with Fedora and your command has looped now over 100 times now without issue.  We did not declare any limits.

**Comment by jackyhallyday** [comment]
I don't know where the bug is, but it only happened with JXL.
I leave this issue open because I can definitely reproduce the bug, and let you decide if it should be closed.

**Comment by urban-warrior** [comment]
JXL is version 0.7.0.  Packages with a version less than 1 suggests the codeset may not be production ready and may have bugs.  What version of JXL are you using (`magick -list format | grep JXL`).

**Comment by jackyhallyday** [comment]
It was already in my previous message  :)

JXL* JXL rw+ JPEG XL (ISO/IEC 18181) (libjxl 0.8.2)
I tested with 1, 2, 4, 6, and 8 threads. Since my CPU has 6 cores/12 threads and runs a webserver and a database too, I stopped there and have set the threads limit to 6.

I usually install ImageMagick with IMEI.
I noted that even the beta that I installed with
```
git clone https://github.com/ImageMagick/ImageMagick.git ImageMagick-7
```
was using libjxl 0.8.2.

If a "normal" ImageMagick installation uses 0.7.0, maybe that's why you don't have that bug?
I was already using IMEI on my previous machine, but it was using libjxl 0.6.1 at that time.
When I migrated to a new server, I installed the latest version, which installed libjxl 0.8.x
